src/main/java/org/opentripplanner/middleware/triptracker/ManageLegTraversal.java

```diff
@@ -0,0 +1,186 @@
+        int minimumSegmentTime = 5;
+        List<Segment> segments = new ArrayList<>();
+        List<Position> positions = new ArrayList<>(orderedPositions);
+        Coordinates groupCoordinates = null;
+        double groupSegmentTime = 0;
+        for (int i=0; i < positions.size()-1; i++) {
+            Coordinates c1 = new Coordinates(positions.get(i).getLatitude(), positions.get(i).getLongitude());
+            Coordinates c2 = new Coordinates(positions.get(i+1).getLatitude(), positions.get(i+1).getLongitude());
+            double timeInSegment = getDistance(c1, c2) * durationPerMeter;
+            if (timeInSegment < minimumSegmentTime) {
+                // Time in segment too small.
+                if (groupCoordinates == null) {
+                    groupCoordinates = c1;
+                    groupSegmentTime = 0;
+                }
+                groupSegmentTime += timeInSegment;
+                if (groupSegmentTime > minimumSegmentTime) {
+                    // Group segment is now big enough.
+                    segments.add(new Segment(groupCoordinates, groupSegmentTime));
+                    groupCoordinates = null;
+                }
+            } else {
+                if (groupCoordinates != null) {
+                    // Group segment is now big enough.
+                    groupSegmentTime += timeInSegment;
+                    segments.add(new Segment(groupCoordinates, groupSegmentTime));
+                    groupCoordinates = null;
+                } else {
+                    segments.add(new Segment(c1, timeInSegment));
+                }
+            }
+        }
+        if (groupCoordinates != null) {
+            // Close incomplete group segment.
+            segments.add(new Segment(groupCoordinates, groupSegmentTime));
+        }
+        return segments;
+    }
+
+    /**
+     * Get the total distance traversed through each position on a leg. This distance is different to that provided
+     * with the leg (distance).
+     */
+    public static double getDistanceTraversedForLeg(SortedSet<Position> orderedPositions) {
+        List<Position> positions = new ArrayList<>(orderedPositions);
+        double total = 0;
+        for (int i=0; i < positions.size()-1; i++) {
+            double dist = getDistance(
+                new Coordinates(positions.get(i).getLatitude(), positions.get(i).getLongitude()),
+                new Coordinates(positions.get(i+1).getLatitude(), positions.get(i+1).getLongitude())
+            );
+            total += dist;
+        }
+        return total;
+    }
+
+    /**
+     * Remove all duplicate points.
+     */
+    public static List<Position> removeDuplicatePoints(List<Position> positions) {
+        return new ArrayList<>(new HashSet<>(positions));
+    }
+
+    /**
+     * Order points to match leg traversal.
+     */
+    public static SortedSet<Position> orderPoints(List<Position> positions) {
```

Comment:
@leonardehrenfried What order do positions come out of `PolylineUtils.decode`? Do they match the order in the encoded polyline? I'm ordering them because once decoded there doesn't seem to be an order applied. This maybe down to this method or more likely the original order in the encoded polyline.

Reply:
Ah, you're using my re-packaged polyline decoder lib! :heart_decoration: 

They should be in the same order as they were encoded. If they aren't then there is a bug somewhere.

What is the source of the polyline here? Is it OTP?

Reply:
@leonardehrenfried The source is indeed OTP.
This is the polyline I'm decoding: [polyline.txt](https://github.com/ibi-group/otp-middleware/files/14167534/polyline.txt)

Reply:
If I use https://geojson.io to decode the polyline, I get one going from South Atlanta towards the city centre and then a bit further north. Is this what you would expect?

Reply:
![Screenshot from 2024-02-05 16-19-13](https://github.com/ibi-group/otp-middleware/assets/151346/06ba2c1d-a435-4115-97f8-285dbc556995)

It's a bit hard to see but it's the black line starting from the airport.

Reply:
Yes, that is the route I'm expecting.

Reply:
I might have been going off on a tangent here, the order from `Polyline.decode` looks ok. Let me get back to you...

Reply:
Points before ordering: 
[points_before_ordering.txt](https://github.com/ibi-group/otp-middleware/files/14168835/points_before_ordering.txt)

Points after ordering:
[points_after_ordering.txt](https://github.com/ibi-group/otp-middleware/files/14168841/points_after_ordering.txt)

There are a lot of differences. I need the points to be ordered so I can traverse the leg.